be/src/exec/pipeline/fragment_context.cpp

```diff
@@ -22,6 +23,53 @@
 
 namespace starrocks::pipeline {
 
+FragmentContext::FragmentContext() : _data_sink(nullptr) {}
+
+FragmentContext::~FragmentContext() {
+    _data_sink.reset();
+    _runtime_filter_hub.close_all_in_filters(_runtime_state.get());
+    clear_all_drivers();
+    close_all_pipelines();
+    if (_plan != nullptr) {
+        _plan->close(_runtime_state.get());
+    }
+}
+
+void FragmentContext::clear_all_drivers() {
+    for (auto& pipe : _pipelines) {
+        pipe->clear_drivers();
+    }
+}
+void FragmentContext::close_all_pipelines() {
+    for (auto& pipe : _pipelines) {
+        pipe->close(_runtime_state.get());
+    }
+}
+
+Status FragmentContext::iterate_drivers(const std::function<Status(const DriverPtr&)>& call) {
+    for (const auto& pipeline : _pipelines) {
+        for (const auto& driver : pipeline->drivers()) {
+            RETURN_IF_ERROR(call(driver));
+        }
+    }
+    return Status::OK();
+}
+
+size_t FragmentContext::total_dop() const {
+    size_t total = 0;
+    for (const auto& pipeline : _pipelines) {
+        total += pipeline->degree_of_parallelism();
+    }
+    return total;
+}
+
+void FragmentContext::set_tplan(TPlan& tplan) {
+    swap(_tplan, tplan);
```

Comment:
That's not intuitive? Why `set_tplan` actually swap with the input?

Reply:
Good advice, let me rename to `move_tplan`
`request.tplan` will be destructed after the RPC request `exec_plan_fragment` finished, so we need move the owner of `tplan` from `request` to `FragmentContext.